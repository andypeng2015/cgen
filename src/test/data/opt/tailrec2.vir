module foo

export function w $foo(w %n, w %a, w %b) {
@start:
  %c = eq.w %n, 0_w
  br %c, @reta, @chkb
@reta:
  ret %a
@chkb:
  %c = eq.w %n, 1_w
  br %c, @retb, @loop
@retb:
  ret %b
@loop:
  %nn = sub.w %n, 1_w
  %na = copy.w %b
  %nb = add.w %a, %b
  %res = call.w $foo(%nn, %na, %nb)
  ret %res
}
