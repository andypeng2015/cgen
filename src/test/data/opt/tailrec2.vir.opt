module foo

export function w $foo(w %n, w %a, w %b) {
@14:
  jmp @2(%n, %a, %b)
@2(%1, %2, %3):
  %8 = eq.w %1, 0x0_w ; @19
  br %8, @13(%2), @4
@4:
  %9 = eq.w %1, 0x1_w ; @20
  br %9, @13(%3), @7
@7:
  %10 = sub.w %1, 0x1_w ; @21
  %11 = add.w %2, %3 ; @22
  jmp @2(%10, %3, %11)
@13(%0):
  ret %0
}
