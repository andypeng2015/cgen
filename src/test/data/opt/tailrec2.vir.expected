module foo

export function w $foo(w %n, w %a, w %b) {
@11:
  jmp @0(%n, %a, %b)
@0(%0, %1, %2):
  %3 = eq.w %0, 0x0_w ; @12
  br %3, @1, @2
@1:
  ret %1
@2:
  %4 = eq.w %0, 0x1_w ; @13
  br %4, @4, @5
@4:
  ret %2
@5:
  %5 = sub.w %0, 0x1_w ; @14
  %6 = add.w %1, %2 ; @15
  jmp @0(%5, %2, %6)
}
